spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultPollableMessageSource.java

```diff
@@ -74,7 +77,9 @@
 
 	private BiConsumer<AttributeAccessor, Message<?>> attributesProvider;
 
-	private volatile boolean running;
+	private SmartMessageConverter messageConverter;
+
+	private boolean running;
```

Comment:
Well, the pattern is still to have it with a`volatile`. Even if you `synchronized` there is still possibility to read from CPU cache, but not jump to the main memory for the actual value. That's exactly why `volatile` has been introduced.
Although with the fixes against Meltdown and Spectre we may not need it anymore because we don't do speculative execution 😄 . Or maybe opposite: since we have done speculative execution before we could read data from other threads. Therefore it is more dangerous do not use `volatile`. Need to think yet...

Just IMHO, but we have convention in Spring Integration do not use `volatile` only for properties which mostly be changed only once during configuration. The multi-thread access must be supplied with `volatile` anyway.
And that's exactly how it isn't good to rely on the `ArrayList.isEmpty()` because its `size` property isn't not `volatile`.

Reply:
No; volatile is not needed when using `synchronized` - synchronization flushes the cache.